### PR TITLE
Added example for client/server communication through websockets

### PR DESCRIPTION
**Pull request by geakstr**

### CODE REVIEW

File: examples/websockets/client.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+var sharedb = require("sharedb/lib/client");
+
+var connection = new sharedb.Connection(new WebSocket("ws://localhost:8888"));
+
+var doc = connection.get("collection", "document");
+
+doc.subscribe(function(error){
+  if (!doc.type) {
```
Comment (geakstr) [inline examples/websockets/client.js]: I'm not sure that this is correct method to verify of doc availability

**Comment by lubaochuan** [comment]
hello geakstr, you example is very helpful. Do you know how to use the client in a browser (like a sharejs client side script)?

**Comment by geakstr** [comment]
@lubaochuan yes, I use sharedb in my project in browser. `client.js` file in this pull request works in browser as well

**Comment by lubaochuan** [comment]
@geakstr Thanks! I used browserify to create a bundle (client.js) of node code for browser use. The client side code runs but it didn't send operations to the server via the websocket correctly (sever.js doesn't print anything when index.html is loaded). Here is my repo for this experiment https://github.com/lubaochuan/sharedb_examples Did I miss something?

**Comment by geakstr** [comment]
@lubaochuan I submit PR into your repo, check this https://github.com/lubaochuan/sharedb_examples/pull/1

File: examples/websockets/server.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+});
+var share = sharedb();
+
+wss.on("connection", function(ws, req) {
+  var stream = new Duplex({
+    objectMode: true
+  });
+
+  stream.headers = ws.upgradeReq.headers;
+  stream.remoteAddress = ws.upgradeReq.connection.remoteAddress;
```
Comment (nornagon) [inline examples/websockets/server.js]: Are these (headers / remoteAddress) necessary for the correct functioning of sharedb? Would be nice to add a comment here if so.

File: examples/websockets/server.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  stream._read = function() {};
+  stream.on("error", function(msg) {
+    ws.close();
+  });
+  stream.on("end", function() {
+    ws.close();
+  });
+
+  ws.on("message", function(op) {
+    // console.log('op', JSON.stringify(JSON.parse(op), null, 2));
```
Comment (nornagon) [inline examples/websockets/server.js]: Can't merge this with random commented-out code lying around :)

File: examples/websockets/client.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    doc.create({x: "hello"});
+  }
+
+  if (doc.type) {
+    // `doc` ready, make ops
+    setTimeout(function() { // timeout need for example (to delay after create)
```
Comment (nornagon) [inline examples/websockets/client.js]: When I tried this, this branch didn't hit in the first connection (the first pageload), so I don't think the timeout is necessary.

File: examples/websockets/client.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+      doc.submitOp([{p: ["x", 5], si: " world"}]);
+    }, 1000);
+  }
+});
```
Comment (nornagon) [inline examples/websockets/client.js]: Would be great to add a listener for `doc.on('op')` to show how that event works.

(Also, mega nitpick: add newlines at the end of the files!)

**Comment by nornagon** [comment]
This was super helpful for getting started with sharedb. Despite being a member of the share org, I haven't used sharejs in years and it's changed a lot. I'd love to have this or something like it in the documentation. @nateps, what do you think about merging this?

**Comment by nateps** [comment]
Hey, Jeremy! Great to hear from you here, and thanks for jumping in with helpful comments.

Having a simple example is a great idea. Thanks, @geakstr!

I think the server code can be simplified a lot. I imagine there should be a library that will create a server-side stream for a websocket connection. Have you tried https://www.npmjs.com/package/websocket-stream?

If we are not using a library, the best way to create a custom stream is typically to inherit from the Duplex class instead of attaching methods to each stream instance. See the implementation of ServerStream in ShareDB: https://github.com/share/sharedb/blob/master/lib/stream-socket.js.

Please add a package.json file in this directory so people can npm install in the example directory and get up and running quickly!

**Comment by geakstr** [comment]
@nateps Hello. I will explore your links a bit later and update pull request. 
@nornagon Thank you for the comments, I will fix it.

**Comment by avital** [comment]
Hi @geakstr, we're aiming to get an simple websocket example ready for the ShareDB 1.0 release, to be released in a few weeks. Do you think you'll have time to respond to the comments? I'd love to use your example.

**Comment by geakstr** [comment]
@nornagon @nateps @avital sorry for delay, guys, I took note of your comments and updated PR

**Comment by geakstr** [comment]
Ah, I see examples folder already exist in repo, unfortunately I did not make it in time :)

**Comment by avital** [comment]
Yeah :)

**Pull request closed without merging**